vita3k/modules/SceHttp/SceHttp.cpp

```diff
@@ -145,57 +504,457 @@ EXPORT(int, sceHttpGetIcmOption) {
+
+    auto req = host.http.requests.find(reqId)->second;
+
+    *header = net_utils::constructHeaders(req.headers);
+
+    *headerSize = strlen(*header) + 1;
+
+    return 0;
+}
+
+EXPORT(int, sceHttpSendRequest, int reqId, const char *postData, size_t size) {
+    if (!host.http.inited)
+        return RET_ERROR(SCE_HTTP_ERROR_BEFORE_INIT);
+
+    if (host.http.requests.find(reqId) == host.http.requests.end())
+        return RET_ERROR(SCE_HTTP_ERROR_INVALID_ID);
+
+    auto req = host.http.requests.find(reqId);
+    auto conn = host.http.connections.find(req->second.connId);
+    auto tmpl = host.http.templates.find(conn->second.tmplId);
+
+    /* TODO:
+        TRACE
+        CONNECT
+     */
+    int bytes, sent, received, total;
+    switch (req->second.method) {
+    case SCE_HTTP_METHOD_GET:
+    case SCE_HTTP_METHOD_HEAD:
+    case SCE_HTTP_METHOD_OPTIONS:
+    case SCE_HTTP_METHOD_DELETE: {
+        auto headers = net_utils::constructHeaders(req->second.headers);
+
+        req->second.message = new char[strlen(req->second.requestLine) + strlen(headers) + 1]();
+        sprintf(req->second.message, "%s\r\n%s\r\n", req->second.requestLine, headers);
+        delete[] headers;
+
+        total = strlen(req->second.message);
+        sent = 0;
+        do {
+            if (conn->second.isSSL) // use SSL_Write if it is an SSL connection
+                bytes = SSL_write((SSL *)tmpl->second.sslConn, req->second.message + sent, total - sent);
+            else
+                bytes = write(conn->second.socketfd, req->second.message + sent, total - sent);
+            if (bytes < 0) {
+                LOG_ERROR("ERROR writing message to socket");
+                return RET_ERROR(SCE_HTTP_ERROR_NETWORK);
+            }
+            if (bytes == 0)
+                break;
+            sent += bytes;
+        } while (sent < total);
+
+        /* receive the response */
+        // The read always separates data and header so its safe to read a bunch a data
+        // and detect whenever we have the full header with \r\n\r\n.
+        auto headersResponse = new char[host.http.defaultResponseHeaderSize]();
+        total = host.http.defaultResponseHeaderSize - 1;
+        received = 0;
+        do {
+            if (conn->second.isSSL) // use SSL_read if it is an SSL connection
+                bytes = SSL_read((SSL *)tmpl->second.sslConn, headersResponse + received, total - received);
+            else
+                bytes = read(conn->second.socketfd, headersResponse + received, total - received);
+            if (bytes < 0) {
+                delete[] headersResponse;
+                LOG_ERROR("ERROR reading response from socket %d", conn->second.socketfd);
+                return RET_ERROR(SCE_HTTP_ERROR_NETWORK);
+            }
+            if (std::string(headersResponse).find("\r\n\r\n") != std::string::npos) {
+                break; // we have  2 line ends, meaning that its the end of the headers
```

Comment:
We need to account for the bytes received in the last chunk of the header (and modify it for other methods too)
```suggestion
               received += bytes;
                break; // we have  2 line ends, meaning that its the end of the headers
```